### PR TITLE
Split members of ClassDef by type

### PR DESCRIPTION
**Pull request by gzm0**
It seems that this is a meaningful distinction for the entire pipeline (even the compiler).

### CODE REVIEW

File: linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala [hunk 1]
```diff
@@ -135,12 +140,12 @@ object TestIRBuilder {
...
-    if (classKind == ClassKind.ModuleClass) List(trivialCtor(className))
-    else if (classKind.isJSClass) List(trivialJSCtor)
-    else Nil
-  }
+  //def requiredMemberDefs(className: ClassName,
+  //    classKind: ClassKind): List[MemberDef] = {
+  //  if (classKind == ClassKind.ModuleClass) List(trivialCtor(className))
+  //  else if (classKind.isJSClass) List(trivialJSCtor)
+  //  else Nil
+  //}
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala:148]: This is a problem :)

**Comment by gzm0** [comment]
Draft for comparison. Interesting is that it seems even most of the compiler gets simplified.

An alternative would be to do this only linker internally (and have a "`SplitClass`" or something that is a `LinkedClass` w/o linking info).

**Comment by gzm0** [comment]
Ah, @sjrd I meant to add you as a reviewer here :shrug:

**Comment by sjrd** [comment]
I've been meaning to look at this as well as #4777 for some days, but couldn't find enough contiguous time to grok them. I'll probably have nothing else to do tomorrow afternoon, though.

File: ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala [hunk 2]
```diff
@@ -1428,243 +1471,232 @@ object Serializers {
...
+          exportedMembersBuilder += exportedMember
       }
+
+      (jsConstructorDefBuilder.result(), exportedMembersBuilder.result())
     }
 
-    def readMemberDef(owner: ClassName, ownerKind: ClassKind): MemberDef = {
-      implicit val pos = readPosition()
-      val tag = readByte()
+    private def readFieldDef()(implicit pos: Position): FieldDef = {
```
Comment (sjrd) [inline ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:1481]: The diff from `readMemberDef` to `readFieldDef+readMethodDef+etc.` is atrocious. :(

Could you perhaps do a first refactoring-only commit that extracts `readFieldDef` and co from the individual `case`s of `readMemberDef`, without touching anything else? The diff for that commit would still be bad, but if it's strictly only moving things in extracted methods, it's easy to review and more importantly easy to ignore in a future `git blame`. Then the second commit will have a nice diff here.

**Review by sjrd — COMMENTED** [review]
Looks nice overall. I like it.

Does it have an impact on #4777? Should this PR be merged first?

File: linker/shared/src/test/scala/org/scalajs/linker/testutils/IRAssertions.scala [hunk 3]
```diff
@@ -78,10 +78,9 @@ object IRAssertions {
   class LinkedClassAssertions(linkedClass: LinkedClass) extends AbstractIRNodeAssertions {
     protected def startTraverse(traverser: Traverser): Unit = {
       linkedClass.jsSuperClass.foreach(traverser.traverse(_))
-      linkedClass.fields.foreach(traverser.traverseMemberDef(_))
```
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/test/scala/org/scalajs/linker/testutils/IRAssertions.scala:81]: Yes. Fields never traversed any Trees and a `Traverser` doesn't have a dedicated method for it ATM. If you want I can add it and traverse.

[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/test/scala/org/scalajs/linker/testutils/IRAssertions.scala:81]: I have added the method to Traverser. Seemed more consistent.

**Pull request merged**